controllers/spacebindingrequestmigration/spacebindingrequest_migration_controller.go

```diff
@@ -0,0 +1,183 @@
+	if err := r.Client.Get(ctx, spaceName, space); err != nil {
+		if errors.IsNotFound(err) {
+			// space was deleted
+			return reconcile.Result{}, nil
+		}
+		// error while reading space
+		return ctrl.Result{}, errs.Wrapf(err, "unable to get the bound Space")
+	}
+
+	murName := types.NamespacedName{Namespace: spaceBinding.Namespace, Name: spaceBinding.Spec.MasterUserRecord}
+	mur := &toolchainv1alpha1.MasterUserRecord{}
+	if err := r.Client.Get(ctx, murName, mur); err != nil {
+		if errors.IsNotFound(err) {
+			// mur was deleted
+			return reconcile.Result{}, nil
+		}
+		// error while reading MUR
+		return ctrl.Result{}, errs.Wrapf(err, "unable to get the bound MUR")
+	}
+
+	// skip workspace creator spacebinding
+	// the controller will convert only spacebindings created by system admins using the sandbox-cli.
+	// If the creator label on the space matches the usersignup reference on the MUR then this is the owner of the space
+	// and the spacebinding should not be migrated.
+	if mur.ObjectMeta.OwnerReferences == nil || len(mur.ObjectMeta.OwnerReferences) == 0 {
+		return ctrl.Result{}, errs.New("MasterUserRecord has no UserSignup owner reference")
+	}
+	usersignup := mur.ObjectMeta.OwnerReferences[0]
+	if space.Labels[toolchainv1alpha1.SpaceCreatorLabelKey] == usersignup.Name {
+		return reconcile.Result{}, nil
+	}
+
+	// get the spaceRole
+	spaceRole := spaceBinding.Spec.SpaceRole
+
+	// get member cluster name where the space was provisioned
+	targetCluster := space.Status.TargetCluster
+	memberCluster, memberClusterFound := r.MemberClusters[targetCluster]
+	if !memberClusterFound {
+		return ctrl.Result{}, errs.New(fmt.Sprintf("unable to find member cluster: %s", targetCluster))
+	}
+
+	// get the home namespace from space
+	defaultNamespace := getDefaultNamespace(space.Status.ProvisionedNamespaces)
+
+	// create the SpaceBindingRequest resource
+	sbrName := mur.GetName() + "-" + spaceRole
+	sbr := &toolchainv1alpha1.SpaceBindingRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      sbrName,
+			Namespace: defaultNamespace,
+			Labels: map[string]string{
+				toolchainv1alpha1.SpaceCreatorLabelKey:                 "spacebindingrequestmigration",
+				toolchainv1alpha1.SpaceBindingSpaceLabelKey:            space.GetName(),
+				toolchainv1alpha1.SpaceBindingMasterUserRecordLabelKey: mur.GetName(),
+			},
+		},
+		Spec: toolchainv1alpha1.SpaceBindingRequestSpec{
+			MasterUserRecord: mur.GetName(),
+			SpaceRole:        spaceRole,
+		},
+	}
+
+	// check if SpaceBindingRequest doesn't exist already
+	err = memberCluster.Client.Get(context.TODO(), types.NamespacedName{
+		Namespace: defaultNamespace,
+		Name:      sbrName,
+	}, &toolchainv1alpha1.SpaceBindingRequest{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// create the sbr resource
+			if err := memberCluster.Client.Create(ctx, sbr); err != nil {
+				return ctrl.Result{}, errs.Wrapf(err, "unable to create SpaceBindingRequest")
+			}
+
+			// delete the SpaceBinding
```

Comment:
There is a race condition. We create an SBR which will trigger a SB creation and we delete the SB here too. We might end up deleting the SB created by the SBR controller. The question is if our SBR controller also watches SBs and will restore the deleted SB. Could you double check that?

Reply:
Initially I was:
- deleting the SpaceBinding
- creating the SpaceBindingRequest

but after discussing this with @MatousJobanek , we decided to invert those, so that we avoid the case in which the SBR is not created and the SpaceBinding was already deleted, which means the user will not have access anymore.

If we go with the current order, could happen that:
- we create a SpaceBindingRequest
- the SBR stays in not provisioned since there is already a SpaceBinding for the same MUR and Space
- the migration controller deletes the existing SpaceBinding
- the SBR controller provisions a new SpaceBinding

> The question is if our SBR controller also watches SBs and will restore the deleted SB. Could you double check that?
yes, this is done here https://github.com/codeready-toolchain/host-operator/blob/567e867d43619a8a7d4f6b63ebf8e0d296be5f38/controllers/spacebindingrequest/spacebindingrequest_controller.go#L46 where we trigger a reconcile on the SpaceBindingRequest if the SpaceBinding is updated/deleted.

Reply:
> - the migration controller deletes the existing SpaceBinding
> - the SBR controller provisions a new SpaceBinding

And in parallel you are deleting the SpaceBinding in this controller. So theoretically it may happen that the SBR controller will ensure the SpaceBinding right before you are deleting it. This is the race condition I'm talking about. But since the SBR controller watches the SB resources it will restore it if it's deleted. So we are good here. Thanks for the confirmation.